azerothcore/azerothcore-wotlk
# (Core/SmartAI): Improve SMART_EVENT_GOSSIP_HELLO

## Changes Proposed:
Add another value to event_param1 for SMART_EVENT_GOSSIP_HELLO to support all cases of GossipHello/reportUse in GameObjects:
- event_param1 set to 0: execute the action for both GossipHello and reportUse. This might result in the action being executed twice when clicking the GameObject
- event_param1 set to 1: execute the action for only GossipHello
- event_param1 set to 2: execute the action for only reportUse

## Issues Addressed:
TC Issue that is replicable on AZ core, that was reported in 2018.
https://github.com/TrinityCore/TrinityCore/issues/21678

## SOURCE:
TC Commit to address the issue that was provided in 2019.
https://github.com/TrinityCore/TrinityCore/commit/9e58ef6940c78d92f1553cf7ab75e87a6e2cf112

## Tests Performed:
Builds without issues.


## How to Test the Changes:

Steps to reproduce the problem before the PR:

1. Execute this sql:
```
UPDATE gameobject_template SET AIName="SmartGameObjectAI" WHERE entry=181147;
DELETE FROM smart_scripts WHERE entryorguid=181147 AND source_type=1;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(181147, 1, 0, 0, 64, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Gob - GOSSIP_HELLO - CLOSE_GOSSIP'),
(181147, 1, 1, 0, 64, 0, 100, 0, 0, 0, 0, 0, 12, 322, 2, 60000, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Gob - GOSSIP_HELLO - SUMMON_CREATURE');
```
2. .Gob add temp 181147
3. Interact with the gob once
4. notice 2 npcs are spawned instead of one

After this PR the issue is no longer a problem.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
